File: web/src/bot.js
// Bot AI: Monte Carlo equity vs random hands, then pot-odds policy.
// Mirrors the Python bot, including difficulty knobs.

import { ALL_CARDS } from "./cards.js";
import { evaluate } from "./evaluator.js";

export const DIFFICULTY_PARAMS = {
  easy:   { sims: 80,  jitter: 0.12, valueThresh: 0.74, callMargin: -0.10, bluffFreq: 0.05, aggr: 0.5 },
  medium: { sims: 220, jitter: 0.05, valueThresh: 0.64, callMargin: 0.00,  bluffFreq: 0.10, aggr: 0.7 },
  hard:   { sims: 420, jitter: 0.02, valueThresh: 0.57, callMargin: 0.04,  bluffFreq: 0.15, aggr: 0.9 },
};

function partialShuffle(deck, need, rng) {
  for (let i = 0; i < need; i++) {
    const j = i + Math.floor(rng() * (deck.length - i));
    const tmp = deck[i]; deck[i] = deck[j]; deck[j] = tmp;
  }
}

export function estimateEquity(hole, community, numOpponents, sims, rng) {
  const known = new Set([...hole, ...community].map(c => c.id));
  const deck = ALL_CARDS.filter(c => !known.has(c.id));
  const needBoard = 5 - community.length;
  const need = numOpponents * 2 + needBoard;

  let score = 0;
  for (let s = 0; s < sims; s++) {
    partialShuffle(deck, need, rng);
    let idx = 0;
    const oppHands = [];
    for (let o = 0; o < numOpponents; o++) { oppHands.push([deck[idx], deck[idx + 1]]); idx += 2; }
    const board = community.concat(deck.slice(idx, idx + needBoard));

    const my = evaluate(hole.concat(board)).score;
    let bestOpp = -1;
    for (const oh of oppHands) {
      const v = evaluate(oh.concat(board)).score;
      if (v > bestOpp) bestOpp = v;
    }
    if (my > bestOpp) score += 1;
    else if (my === bestOpp) {
      let ties = 0;
      for (const oh of oppHands) if (evaluate(oh.concat(board)).score === my) ties++;
      score += 1 / (ties + 1);
    }
  }
  return score / sims;
}

// ctx: { community, currentBet, minRaise, pot, numOpponents, bigBlind, rng, street }
// Returns [action, amount]; for "raise", amount is the total street bet to raise to.
export function decide(player, ctx) {
  const p = DIFFICULTY_PARAMS[player.difficulty];
  const rng = ctx.rng;

  let equity = estimateEquity(player.hole, ctx.community, ctx.numOpponents, p.sims, rng);
  equity += (rng() * 2 - 1) * p.jitter;
  equity = Math.max(0, Math.min(1, equity));

  const toCall = ctx.currentBet - player.streetBet;

  const raiseTo = () => {
    const add = Math.max(ctx.minRaise, Math.floor(ctx.pot * p.aggr));
    const target = ctx.currentBet + add;
    const maxTarget = player.streetBet + player.chips; // all-in shove
    return Math.min(target, maxTarget);
  };

  // No bet to call: check or bet.
  if (toCall <= 0) {
    if (equity >= p.valueThresh && player.chips > 0) return ["raise", raiseTo()];
    if (rng() < p.bluffFreq && player.chips > ctx.bigBlind) return ["raise", raiseTo()];
    return ["check", 0];
  }

  // Facing a bet.
  const potOdds = toCall / (ctx.pot + toCall);

  if (equity >= p.valueThresh && player.chips > 0) {
    if ((ctx.street === "turn" || ctx.street === "river") && rng() < 0.25) return ["call", 0];
    return ["raise", raiseTo()];
  }
  if (equity >= potOdds + p.callMargin) return ["call", 0];
  if (rng() < p.bluffFreq && toCall <= ctx.pot * 0.5 && player.chips > toCall * 2) {
    return ["raise", raiseTo()];
  }
  return ["fold", 0];
}

File: bot.py
"""Bot decision making.

Strategy: estimate win probability (equity) by Monte Carlo simulation against
random opponent holdings, then act on pot odds with a difficulty-tuned policy.
All three difficulties share this core; they differ in simulation accuracy,
discipline, aggression and bluff frequency.
"""

from __future__ import annotations

import random
from dataclasses import dataclass

from cards import Card, RANKS, SUITS
from evaluator import evaluate
from player import Player

# Per-difficulty knobs.
#   sims        : Monte Carlo samples (more = more accurate equity)
#   jitter      : random noise added to equity (more = sloppier reads)
#   value_thresh: equity at/above which the bot raises for value
#   call_margin : required equity above pot odds to call (negative = loose)
#   bluff_freq  : chance to bluff/semi-bluff in a given spot
#   aggr        : raise size as a fraction of the pot
DIFFICULTY_PARAMS = {
    "easy":   dict(sims=80,  jitter=0.12, value_thresh=0.74, call_margin=-0.10, bluff_freq=0.05, aggr=0.5),
    "medium": dict(sims=220, jitter=0.05, value_thresh=0.64, call_margin=0.00,  bluff_freq=0.10, aggr=0.7),
    "hard":   dict(sims=420, jitter=0.02, value_thresh=0.57, call_margin=0.04,  bluff_freq=0.15, aggr=0.9),
}


@dataclass
class Context:
    community: list[Card]
    current_bet: int     # highest street_bet at the table this round
    min_raise: int       # minimum legal raise increment
    pot: int             # chips in the pot (already committed by everyone)
    num_opponents: int   # opponents still in the hand (not folded), >= 1
    big_blind: int
    rng: random.Random
    street: str          # "preflop" | "flop" | "turn" | "river"


def estimate_equity(hole: list[Card], community: list[Card],
                    num_opponents: int, sims: int, rng: random.Random) -> float:
    """Monte Carlo win probability vs `num_opponents` random hands."""
    known = set(hole) | set(community)
    deck = [Card(r, s) for s in SUITS for r in RANKS if Card(r, s) not in known]

    need_board = 5 - len(community)
    need = num_opponents * 2 + need_board

    score = 0.0
    for _ in range(sims):
        sample = rng.sample(deck, need)
        idx = 0
        opp_hands = []
        for _ in range(num_opponents):
            opp_hands.append(sample[idx:idx + 2])
            idx += 2
        board = community + sample[idx:idx + need_board]

        my = evaluate(hole + board).key
        best_opp = max(evaluate(oh + board).key for oh in opp_hands)
        if my > best_opp:
            score += 1.0
        elif my == best_opp:
            # Count how many opponents tie us for an even split.
            ties = sum(1 for oh in opp_hands if evaluate(oh + board).key == my)
            score += 1.0 / (ties + 1)
    return score / sims


def decide(player: Player, ctx: Context) -> tuple[str, int]:
    """Return (action, amount). For "raise", amount is the total street bet to
    raise to; otherwise amount is 0."""
    p = DIFFICULTY_PARAMS[player.difficulty]
    rng = ctx.rng

    equity = estimate_equity(player.hole, ctx.community, ctx.num_opponents,
                             p["sims"], rng)
    # Sloppier bots misread their equity.
    equity += rng.uniform(-p["jitter"], p["jitter"])
    equity = max(0.0, min(1.0, equity))

    to_call = ctx.current_bet - player.street_bet

    def raise_to() -> int:
        add = max(ctx.min_raise, int(ctx.pot * p["aggr"]))
        target = ctx.current_bet + add
        # Cap at an all-in shove.
        max_target = player.street_bet + player.chips
        return min(target, max_target)

    # --- No bet to call: we may check or bet. ---
    if to_call <= 0:
        if equity >= p["value_thresh"] and player.chips > 0:
            return ("raise", raise_to())
        if rng.random() < p["bluff_freq"] and player.chips > ctx.big_blind:
            return ("raise", raise_to())
        return ("check", 0)

    # --- Facing a bet. ---
    pot_odds = to_call / (ctx.pot + to_call)

    # Strong: raise for value (occasionally just call to trap on later streets).
    if equity >= p["value_thresh"] and player.chips > 0:
        if ctx.street in ("turn", "river") and rng.random() < 0.25:
            return ("call", 0)
        return ("raise", raise_to())

    # Priced in: call.
    if equity >= pot_odds + p["call_margin"]:
        return ("call", 0)

    # Semi-bluff: occasionally raise a missed-but-cheap spot.
    if (rng.random() < p["bluff_freq"]
            and to_call <= ctx.pot * 0.5
            and player.chips > to_call * 2):
        return ("raise", raise_to())

    return ("fold", 0)
